<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
	<head>
		<meta http-equiv="Content-Type" content="text/html; charset=utf-8" /><title>
		</title>
		<style type="text/css">
			.cs95E872D0{text-align:left;text-indent:0pt;margin:0pt 0pt 0pt 0pt}
			.csAC91E297{color:#000000;background-color:#FFFFFF;font-family:Book Antiqua;font-size:14pt;font-weight:bold;font-style:normal;}
			.cs57154F9C{color:#000000;background-color:transparent;font-family:Book Antiqua;font-size:12pt;font-weight:normal;font-style:normal;}
		</style>
	</head>
	<body>
		<p class="cs95E872D0"><span class="csAC91E297">Java - Exceptions</span></p><p class="cs95E872D0"><span class="cs57154F9C">&nbsp;</span></p><p class="cs95E872D0"><span class="cs57154F9C">An exception is an event that occurs during the execution(or run time) of a program that disrupts(stop) the normal flow of control.<br/><br/>If an exception is raised( at run time ), the program terminates by displaying the cause of exception. If we handle the exception properly, the program flow will be normal and continues further. This facility is not available in C language.&nbsp;<br/><br/>All exceptions can be broadly divided into two categories:-<br/>1. Checked exceptions&nbsp;<br/>2. Unchecked exceptions.<br/></span></p><p class="cs95E872D0"><span class="cs57154F9C"><br/>Unchecked exceptions:-<br/>All the subclasses of RuntimeException are called unchecked exceptions. Some examples are ArithmeticException and ArrayIndexOutOfBoundsException. These exceptions are called unchecked for the reason, even if the programmer does not provide any try - catch mechanism, the program compiles.&nbsp;<br/><br/>System checks them at runtime and if an exception occurs program terminates. If a try - catch block is provided for these exceptions, program execution continues further when exception really occurs.<br/></span></p><p class="cs95E872D0"><span class="cs57154F9C"><br/>Checked exceptions:-<br/>All the subclasses of Exception class except RuntimeException are called checked exceptions. For these exceptions, programmer should provide a try - catch block whether really exception raises or not, else program does not compile.&nbsp;<br/><br/>Generally, these exceptions are written in the method signature itself with throws clause. Following are the examples from Java API:&nbsp;<br/><br/>public FileInputStream( String filename ) throws FileNotFoundException<br/>public void sleep( int milliseconds ) throws InterruptedException&nbsp;<br/><br/>In the above statements, FileNotFoundException and InterruptedException are checked exceptions. That is, when FileInputStream( String filename ) constructor is used by the programmer, he should write a try - catch mechanism for FileNotFoundException whether the file passed as parameter really exists in the hard disk or not.&nbsp;<br/><br/>Note:&nbsp;Write the following program in the examination for explaining try, catch and finally blocks.</span></p><p class="cs95E872D0"><span class="cs57154F9C">&nbsp;</span></p><p class="cs95E872D0"><span class="cs57154F9C">public class ExceptionHandled &nbsp;&nbsp;&nbsp;</span></p><p class="cs95E872D0"><span class="cs57154F9C">{</span></p><p class="cs95E872D0"><span class="cs57154F9C">	public static void main(String args[ ] ) &nbsp;&nbsp;</span></p><p class="cs95E872D0"><span class="cs57154F9C">	{</span></p><p class="cs95E872D0"><span class="cs57154F9C">	 &nbsp;int a = 5, b = 0, c ;</span></p><p class="cs95E872D0"><span class="cs57154F9C">		 </span></p><p class="cs95E872D0"><span class="cs57154F9C">	 &nbsp;&nbsp;try </span></p><p class="cs95E872D0"><span class="cs57154F9C">	 &nbsp;&nbsp;{</span></p><p class="cs95E872D0"><span class="cs57154F9C">		 c = a / b ;</span></p><p class="cs95E872D0"><span class="cs57154F9C">	 &nbsp;&nbsp;&nbsp;&nbsp;System.out.println( c ) ;</span></p><p class="cs95E872D0"><span class="cs57154F9C">					</span></p><p class="cs95E872D0"><span class="cs57154F9C">		 &nbsp;}</span></p><p class="cs95E872D0"><span class="cs57154F9C">	 &nbsp;&nbsp;&nbsp;catch( ArithmeticException e ) &nbsp;</span></p><p class="cs95E872D0"><span class="cs57154F9C">	 &nbsp;&nbsp;&nbsp;&nbsp;{</span></p><p class="cs95E872D0"><span class="cs57154F9C">		 &nbsp;System.out.println( &quot;Don&#39;t divide by zero sir &quot; + e ) ;</span></p><p class="cs95E872D0"><span class="cs57154F9C">			}</span></p><p class="cs95E872D0"><span class="cs57154F9C">		finally </span></p><p class="cs95E872D0"><span class="cs57154F9C">		 {</span></p><p class="cs95E872D0"><span class="cs57154F9C">		 &nbsp;&nbsp;System.out.println( &quot;Guaranteed execution &quot; ) ;</span></p><p class="cs95E872D0"><span class="cs57154F9C">		 &nbsp;&nbsp;}</span></p><p class="cs95E872D0"><span class="cs57154F9C">		}</span></p><p class="cs95E872D0"><span class="cs57154F9C">	}</span></p><p class="cs95E872D0"><span class="cs57154F9C">&nbsp;</span></p><p class="cs95E872D0"><span class="cs57154F9C">&nbsp;</span></p><p class="cs95E872D0"><span class="cs57154F9C">try :<br/>try is a keyword in Java that defines a block of statements. First, we must identify the statements, in a program, that may raise the exception. Then place those statements in the try block.&nbsp;<br/><br/>That is, the statements that are placed in the try block, may cause( throw ) the exception. If exception is not raised, the system simply ignores the try block.<br/><br/>catch:<br/>catch is a keyword in Java and denotes a block of statements. Every try block should be followed by a catch block. The exception thrown by the try block is caught( or handled ) by the catch block.&nbsp;<br/><br/>We must provide a suitable exception class that can handle( or catch) the exception thrown by the try block. That is, catch block contains the exception handler.&nbsp;<br/><br/>When an exception handler present in the catch block successfully catches the exception, the program execution goes in a normal way and continues further.&nbsp;<br/><br/>One program can contain any number of try - catch blocks. A try block can have any number of catch blocks.&nbsp;<br/></span></p><p class="cs95E872D0"><span class="cs57154F9C"><br/>finally:<br/>finally is a keyword in Java and denotes a block of statements. The statements placed in the finally block are definitely executed whether catch block successfully catches the exception or not.<br/><br/>Generally, we write clean up statements in this block, like closing of streams or closing of sockets etc. finally block is optional.&nbsp;</span></p><p class="cs95E872D0"><span class="cs57154F9C">&nbsp;</span></p></body>
</html>
